pulsar-client/src/main/java/org/apache/pulsar/client/impl/PulsarChannelInitializer.java

```diff
@@ -209,5 +209,33 @@ CompletableFuture<Channel> initSocks5IfConfig(Channel ch) {
 
         return initSocks5Future;
     }
+
+    CompletableFuture<Channel> initializeClientCnx(Channel ch,
+                                                   InetSocketAddress logicalAddress,
+                                                   InetSocketAddress resolvedPhysicalAddress) {
+        CompletableFuture<Channel> initFuture = new CompletableFuture<>();
+        ch.eventLoop().execute(() -> {
+            final ClientCnx cnx = (ClientCnx) ch.pipeline().get("handler");
+
+            if (cnx == null) {
+                initFuture
+                        .completeExceptionally(new IllegalStateException("Missing ClientCnx. This should not happen."));
+            }
+
+            // Need to do our own equality because the physical address is resolved already
+            if (!(logicalAddress.getHostString().equalsIgnoreCase(resolvedPhysicalAddress.getHostString())
+                    && logicalAddress.getPort() == resolvedPhysicalAddress.getPort())) {
+                // We are connecting through a proxy. We need to set the target broker in the ClientCnx object so that
+                // it can be specified when sending the CommandConnect.
+                cnx.setTargetBroker(logicalAddress);
+            }
+
+            cnx.setRemoteHostName(resolvedPhysicalAddress.getHostString());
+
+            initFuture.complete(ch);
+        });
```

Comment:
instead of handling the CompletableFuture logic manually, it might be good to use `NettyFutureUtil.toCompletableFuture(ch.eventLoop().submit(() -> {...}, ch))`.

The benefit of this is that all exceptions would be caught.

Reply:
I didn't know about that one. That seems helpful. I'll look into it.